src/relax/transform/dataflow_inplace.cc

```diff
@@ -0,0 +1,1017 @@
+        // can't be done in-place, ignore
+        if (target_sinfo.empty()) {
+          continue;
+        }
+
+        // Check that at least one argument matches size with the result
+        for (size_t j = 0; j < call_node->args.size(); j++) {
+          auto arg = call_node->args[j];
+          for (auto target : target_sinfo) {
+            std::pair<bool, bool> match = SizeMatches(target, GetStructInfo(arg));
+            if (match.first) {
+              candidates.insert(static_cast<int>(j));
+              if (match.second) {
+                exact_match_candidates.insert(static_cast<int>(j));
+              }
+            }
+          }
+        }
+        if (!candidates.size()) {
+          continue;
+        }
+
+        // Make sure at least one candidate is not live past this point and does not have an alias
+        // live past this point
+        std::unordered_set<int> remove_candidates;
+        for (auto candidate : candidates) {
+          if (!InplaceConditionsMet(live_ranges, alias_sets, tuple_map, currently_live,
+                                    call_node->args[candidate], i)) {
+            remove_candidates.insert(candidate);
+          }
+        }
+        // (remove now to avoid modifying the list as we iterate on it)
+        for (auto candidate : remove_candidates) {
+          candidates.erase(candidate);
+        }
+
+        // if we have a candidate, then this can be made in-place. Report the appropriate candidates
+        if (!candidates.size()) {
+          continue;
+        }
+
+        // produce a list of candidates for this index
+        std::vector<int> size_match_indices = {static_cast<int>(i)};
+        size_match_indices.insert(size_match_indices.end(), candidates.begin(), candidates.end());
+        size_match_list.push_back(size_match_indices);
+
+        // also gather up the exact match candidates if there are any
+        std::vector<int> exact_match_indices = {static_cast<int>(i)};
+        for (auto candidate : candidates) {
+          if (exact_match_candidates.count(candidate)) {
+            exact_match_indices.push_back(candidate);
+          }
+        }
+        if (exact_match_indices.size() > 1) {
+          exact_match_list.push_back(exact_match_indices);
+        }
+      }
+    }
+
+    // remove vars whose range has come to an end
+    // (keep a separate set to avoid changing the sit while iterating on it)
+    std::unordered_set<Var, ObjectPtrHash, ObjectPtrEqual> remove;
+    for (auto var : currently_live) {
+      auto live_range = live_ranges[var];
+      if (live_range.second <= static_cast<int>(i)) {
+        remove.insert(var);
+      }
+    }
+    for (auto var : remove) {
+      currently_live.erase(var);
+    }
+  }
+
+  return {size_match_list, exact_match_list};
+}
+
+// Replace buffers in a PrimFunc according to the mapping.
+tir::Stmt RemapBuffers(const tir::Stmt& stmt, const Map<tir::Buffer, tir::Buffer>& buffer_map) {
```

Comment:
I think this could be handled by calling [`tir::Specialize`](https://github.com/apache/tvm/blob/main/include/tvm/tir/function.h#L267).

Reply:
Ooh! I didn't know about it

Reply:
It's a nice one for a lot of signature manipulations.  There may need to be some post-processing, but it should work in this case.  If the selected in-place input is parameter `i`, and the output is parameter `j`, then it would be called as `auto new_func = Specialize(func, {{func->params[j], func->buffer_map[func->params[i]]}})`.

Reply:
Hm, using `tir::Specialize` still left some uses of the old output buffer around, which still seems like it would require the manual rewrite I had. It was odd. Namely, the old buffer was being used in the `writes` field of the block.

Reply:
There was no DeclBuffer, there was just a loose variable hanging around.

Reply:
Got it.  In that case, it would fall into the same bug that #14565 was intended to resolve.  I think for now, it's probably better to use the separate mutator.  If either RFC#70 gets fully implemented, or if the relax-to-TIR conversion provides a `DeclBuffer` annotation, then it could be updated.